src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -726,14 +750,90 @@ def _visit_deref(self, node: itir.FunCall) -> list[ValueExpr]:
 
             return [ValueExpr(value=result_access, dtype=iterator.dtype)]
 
-        else:
+        elif all([dim in iterator.indices for dim in iterator.dimensions]):
+            # The deref iterator has index values on all dimensions: the result will be a scalar
             args = [ValueExpr(iterator.field, iterator.dtype)] + [
-                ValueExpr(iterator.indices[dim], iterator.dtype) for dim in sorted_dims
+                ValueExpr(iterator.indices[dim], _INDEX_DTYPE) for dim in sorted_dims
             ]
             internals = [f"{arg.value.data}_v" for arg in args]
             expr = f"{internals[0]}[{', '.join(internals[1:])}]"
             return self.add_expr_tasklet(list(zip(args, internals)), expr, iterator.dtype, "deref")
 
+        else:
+            # Not all dimensions are included in the deref index list:
+            # this means the ND-field will be sliced along one or more dimensions and the result will be an array
+            field_array = self.context.body.arrays[iterator.field.data]
+            result_shape = tuple(
+                dim_size
+                for dim, dim_size in zip(sorted_dims, field_array.shape)
+                if dim not in iterator.indices
+            )
+            result_name = unique_var_name()
+            self.context.body.add_array(result_name, result_shape, iterator.dtype, transient=True)
+            result_array = self.context.body.arrays[result_name]
+            result_node = self.context.state.add_access(result_name)
+
+            deref_connectors = ["_inp"] + [
+                f"_i_{dim}" for dim in sorted_dims if dim in iterator.indices
+            ]
+            deref_nodes = [iterator.field] + [
```

Comment:
The solution proposed here is to build a nested sdfg and map the index values to sdfg symbols. This allows inside the nested sdfg to slice the source field by specifying the subset in a memlet copy. The reason why we need a nested sdfg is because the memlet cannot access the scalar values directly, from the index nodes.

This solution is inefficient, unfortunately. It will probably be impossible for DaCe to inline this sdfg. Another solution which was considered is to use a map and copy the requested slice element-by-element from the source array. This solution does not require a nested sdfg, however it will probably be impossible for DaCe to optimize away the map and instead generate memory-copy along each slice dimension.

Having to decide between sdfg nesting and element-by-element copy, I chose the first one. SDFG nesting results in one additional level of function call, but it should preserve memory-copy in each slice dimension, at least from a representation point-of-view. From the dace templates used for code generation, it looks like the dynamic copy will anyway result in an element-by-element copy.
```
inline void deref_1_0_83(my_stencil_state_t *__state, const long long&  _i_Edge, double* __restrict__ _inp, double* __restrict__ _out, long long __var_272__shp, long long __var_274__strd, long long __var_275__strd) {
    long long _sym_i_Edge;

    _sym_i_Edge = _i_Edge;
    {
        dace::CopyNDDynamic<double, 1, false, 1>::template ConstDst<1>::Copy(
        _inp + (__var_275__strd * _sym_i_Edge), _out, __var_272__shp, __var_274__strd);
    }
}
```

Reply:
Just for a reference:
- My [Gather](https://github.com/philip-paul-mueller/jax_to_dace/blob/philip.dev/JaxprToSDFG/translators/gatherTranslator.py) instruction that basically performs indirection, i.e. reads indexes from an array and performs the accesses in another array.
- My [Slicer](https://github.com/philip-paul-mueller/jax_to_dace/blob/c5db346ae9793bb226cd5e1ec5e4bab79de4bcf6/JaxprToSDFG/translators/slicingTranslator.py#L115).

The code above assumes fixed sizes at compile time, which is simpler.